### PR TITLE
chore(unit_test,userspace): allow env var to get expanded in yaml even when part of a string

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area engine

**What this PR does / why we need it**:

Moreover, support env variable embedding another env variable.

Basically, this PR would allow us to use eg:
```
engine:
  ebpf:
    probe: /${HOME}/.falco/falco-bpf.o
```
in #2413.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new(userspace): support env variable expansion in all yaml, even inside strings.
```

### CODE REVIEW

**Comment by FedeDP** [comment]
Expands https://github.com/falcosecurity/falco/pull/2562.

File: unit_tests/falco/test_configuration.cpp
Comment (FedeDP) [inline unit_tests/falco/test_configuration.cpp:139]: Tests were expanded:
* new case with whitespace AFTER the expanded string. Moreover, since we now support env variables embedded inside bigger strings, both ` ${ENV_VAR}` and `${ENV_VAR} ` are now expanded
* we do now tests that env variables found in the middle of strings are properly expanded (`paths[0,1,2,3]` cases)
* `paths[4]` checks an even weirder case, ie: when the env variable gets expanded to another env variable. It must get resolved twice.

File: userspace/falco/yaml_helper.h
Comment (FedeDP) [inline userspace/falco/yaml_helper.h:85]: Short-circuite, when empty value is passed, just return default.

File: userspace/falco/yaml_helper.h
Comment (FedeDP) [inline userspace/falco/yaml_helper.h:92]: For strings, just return it.
This workarounds the fact that `stringstream` splits values by whitespaces, and this is something we don't want to be done when converting string->string.

File: userspace/falco/yaml_helper.h
Comment (FedeDP) [inline userspace/falco/yaml_helper.h:96]: Enforce `std::boolalpha` so that `true` and `false` are correctly parsed to booleans.

**Comment by FedeDP** [comment]
/cc @jasondellaluce 

I am not sure whether we want this for 0.37.0, leave it to other maintainers to decide ;) 
Also cc @therealdwright who first opened the original PR!

File: userspace/falco/yaml_helper.h
Comment (incertum) [reply userspace/falco/yaml_helper.h:85]: Do we have a test for it? I may have overlooked it.

**Review by incertum — APPROVED** [review]
/approve

File: unit_tests/falco/test_configuration.cpp
Comment (FedeDP) [reply unit_tests/falco/test_configuration.cpp:235]: On windows we should map `unsetenv` to  `_putenv(string,"")` anyway, so i decided to avoid adding a new macro and just use what i got :)

File: unit_tests/falco/test_configuration.cpp
Comment (FedeDP) [reply unit_tests/falco/test_configuration.cpp:201]: Thanks :D

File: userspace/falco/yaml_helper.h
Comment (FedeDP) [reply userspace/falco/yaml_helper.h:85]: Will double check and eventually add one!

**Comment by FedeDP** [comment]
> Do we have a test for it? I may have overlooked it.


Added in latest commit, thank you!

**Review by incertum — APPROVED** [review]
/approve

**Comment by FedeDP** [comment]
Trying to put this in the 0.37.0 milestone; let's see if we can make this happen :D 
/milestone 0.37.0

File: unit_tests/falco/test_configuration.cpp
Comment (FedeDP) [reply unit_tests/falco/test_configuration.cpp:151]: We check that `plugins` config (that is getted as a sequence)  is still correctly expanded.

File: userspace/falco/yaml_helper.h
Comment (FedeDP) [inline userspace/falco/yaml_helper.h:45]: `yaml_visitor` is a small visitor API for yaml files, to call a callback function on each scalar value.
It is private to `yaml_helper` since it receives non-constant `YAML::Node`s (thus it can modify the parsed yaml structure).

File: userspace/falco/yaml_helper.h
Comment (FedeDP) [inline userspace/falco/yaml_helper.h:119]: No more custom logic needed.

Comment (FedeDP) [reply userspace/falco/yaml_helper.h:119]: BREAKING CHANGE: in https://github.com/falcosecurity/falco/pull/2562 env var that resolved to empty string (both non-set env vars and env vars set to "") returned default value.
This was in constrast with everything else in our yaml impl (ie: we only returned default value if node is not defined).

Comment (FedeDP) [reply userspace/falco/yaml_helper.h:119]: BREAKING CHANGE: `node.as<T>(default_value);` will return `default_value` when it is not able to parse the node to `T`.

**Comment by FedeDP** [comment]
/unhold

Now we support env vars expansion in all getters, since we `pre-process` the yaml root to expand any scalar value as soon as the document is loaded.

**Comment by leogr** [comment]
/assign

I'll take a look as soon as I can

**Review by leogr — APPROVED** [review]
As discussed with @FedeDP this PR introduces a minor behavior change that needs to be signaled and documented.


Anyway, the feature is great and the PR LGTM.

@incertum PTAL again

**Review by incertum — APPROVED** [review]
/approve

**Pull request merged**
